pkgs/applications/window-managers/picom/picom-allusive.nix

```
@@ -0,0 +1,24 @@
+{ picom, lib, fetchFromGitHub }:
+
+picom.overrideAttrs (oldAttrs: rec {
+  pname = "picom-allusive";
+  version = "0.3.1";
+
+  src = fetchFromGitHub {
+    owner = "allusive-dev";
+    repo = "picom-allusive";
+    rev = version;
+    hash = "sha256-lk4Ll0mi9h3BAqwgOzFQw4WYKnSW9XTl3PjoK2E4WKg=";
+  };
+
+  postInstall = ''
+    chmod +x $out/bin/picom-trans
+  '' + oldAttrs.postInstall;
```

To prevent `postInstall` being removed from the overridden package breaking the eval:

```suggestion
  '' + (lib.optionalString (oldAttrs ? postInstall) oldAttrs.postInstall);
```

I would usually prefer/use a set-pattern argument (`{ postInstall ? "", ... }: {/*...*/}`) for this purpose, but `rec` is wrecking (heh) the scope here, meaning `postInstall` would infrec on the attrset's own `postInstall`.

---

I have no idea what you are saying. Can you please explain it more simply?

---

Sure!

* * *

### About `overrideAttrs`

The `overrideAttrs` function takes one argument; that argument is a function, which receives an *attribute set* as a parameter; the attribute set (if you allow the oversimplification) is all attributes provided to `mkDerivation` in the original derivation.

Here's how it looks like with `builtins.trace` in the repl, doing a "no-op" to `hello`.

```
nix-repl> hello.overrideAttrs (x: builtins.trace x x)
trace: { doCheck = true; meta = <CODE>; passthru = <CODE>; pname = "hello"; src = <CODE>; version = "2.12.1"; }
«derivation /nix/store/vz9naq2p06k5sagqz5mxcbmff0a4d3hb-hello-2.12.1.drv»
```

 

In your particular case, `oldAttrs` is the argument name to the function passed to `overrideAttrs`:

```text
picom.overrideAttrs (oldAttrs: rec {
                     ^^^^^^^^ This is the argument to the function given to `overrideAttrs`
```

And in turn, when referring to `oldAttrs.installPhase`, it refers to this attribute:

 - https://github.com/NixOS/nixpkgs/blob/5a9c737c587d2c34d63c5b3cb53c6ab0705bdf4f/pkgs/applications/window-managers/picom/default.nix#L89-L94

*if* this attribute ends-up being removed through a refactor of the `picom` package, the evaluation will fail; here's how it would fail, using `hello`, which doesn't have a `postInstall`:

```
nix-repl> hello.overrideAttrs (oldAttrs: { postInstall = "" + oldAttrs.postInstall; })
       error: attribute 'postInstall' missing

       at «string»:1:53:

            1| hello.overrideAttrs (oldAttrs: { postInstall = "" + oldAttrs.postInstall; })
             |                                                     ^
```

We can guard against this problem like so:

```
nix-repl> hello.overrideAttrs (oldAttrs: { postInstall = "" + (lib.optionalString (oldAttrs ? postInstall) oldAttrs.postInstall); })  
«derivation /nix/store/bvw2jx898i33dzih5zn4bwh499d7vns7-hello-2.12.1.drv»
```

Please keep on reading :)

* * *

### `a ? x` ? ??

 - https://nixos.org/manual/nix/stable/language/operators#has-attribute

Using the `?` operator we're checking the attribute set has the given attribute.

```
nix-repl> { a = 1; } ? a
true

nix-repl> { b = 1; } ? a 
false
```

* * *

### A set-pattern what now?

Second list item of the *Functions* part of the language constructs page:

 - https://nixos.org/manual/nix/stable/language/constructs#functions

A *set pattern argument* is a special kind of argument to a function that is close to what *destructuring* is in other languages.

Now, using our contrived example with `hello` in mind, we can instead do this:

```
nix-repl> hello.overrideAttrs ({ postInstall ? "", ... }: { postInstall = "" + postInstall; })                                       
«derivation /nix/store/bvw2jx898i33dzih5zn4bwh499d7vns7-hello-2.12.1.drv»
```

*But why can't we do this here?* As I alluded to, it's `rec` w`rec`king our plans.

```
nix-repl> hello.overrideAttrs ({ postInstall ? "", ... }: rec { postInstall = "" + postInstall; })
error:
...
       error: infinite recursion encountered

       at «string»:1:74:

            1| hello.overrideAttrs ({ postInstall ? "", ... }: rec { postInstall = "" + postInstall; })
             |                                                                          ^
```

`rec` makes the attribute set [a recursive set](https://nixos.org/manual/nix/stable/language/constructs#recursive-sets), which in turn, means that the attribute set's *own* attributes are now in scope. So the bare `postInstall` in the attribute set refers to the value it is trying to set, causing the infinite recursion.

In fact, this is entirely independent to the `overrideAttrs` function call!

```
nix-repl> rec { postInstall = "" + postInstall; }                                                  
{ postInstall = error: infinite recursion encountered

       at «string»:1:26:

            1| rec { postInstall = "" + postInstall; }
             |                          ^
```

* * *

Does that help you understand what I meant? Don't worry if you have more questions, do ask.

---

Seems quite complex but I think I understand. I'll give it a try now.